packages/backend-common/src/service/lib/ServiceBuilderImpl.ts

```diff
@@ -14,44 +14,74 @@
  * limitations under the License.
  */
 
+import { ConfigReader } from '@backstage/config';
 import compression from 'compression';
 import cors from 'cors';
 import express, { Router } from 'express';
 import helmet from 'helmet';
 import { Server } from 'http';
 import stoppable from 'stoppable';
 import { Logger } from 'winston';
-import { getRootLogger } from '../logging';
+import { useHotCleanup } from '../../hot';
+import { getRootLogger } from '../../logging';
 import {
   errorHandler,
   notFoundHandler,
   requestLoggingHandler,
-} from '../middleware';
-import { ServiceBuilder } from './types';
-import { useHotCleanup } from '../hot';
+} from '../../middleware';
+import { ServiceBuilder } from '../types';
+import { readBaseOptions, readCorsOptions } from './config';
 
 const DEFAULT_PORT = 7000;
+const DEFAULT_HOST = 'localhost';
 
 export class ServiceBuilderImpl implements ServiceBuilder {
   private port: number | undefined;
+  private host: string | undefined;
   private logger: Logger | undefined;
   private corsOptions: cors.CorsOptions | undefined;
   private routers: [string, Router][];
-  /**
-   * Reference to the module where builder is created
-   * Needed for the HMR
-   */
+  // Reference to the module where builder is created - needed for hot module
+  // reloading
   private module: NodeModule;
+
   constructor(module: NodeModule) {
     this.routers = [];
     this.module = module;
   }
 
+  loadConfig(config: ConfigReader): ServiceBuilder {
+    const backendConfig = config.getOptionalConfig('backend');
+    if (!backendConfig) {
+      return this;
+    }
+
+    const baseOptions = readBaseOptions(backendConfig);
+    if (baseOptions.bindPort) {
+      this.port = baseOptions.bindPort;
```

Comment:
Could've just `this.port = backendConfig.getOptionalNumber('backend.bindPort') ?? DEFAULT_PORT` here right?

Reply:
No. The default port shall be applied even if the config read is never triggered

Reply:
Huh? You mean you want `this.port = backendConfig.getOptionalNumber('backend.bindPort') ?? this.port`?

Reply:
Hm i just mean that if we only do the default fallback in `private getOptions`, the code will work no matter if the user called setPort instead of loading config etc

Reply:
Actually can we kill this discussion and just go with the `listen` option instead? :grin: Then it's more of a straightforward `const listen = config.getOptionalStringArray('listen'); if (listen) ...`